# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsGlobalWindowOuter.cpp
# Commit: ae01b3db6250
# Full Hash: ae01b3db6250b366ac975c4257a384c1de13d632
# Author: Dimi Lee <dlee@mozilla.com>
# Date: 2020-04-23 21:43:09
# Regressor Bug: 1616775
# File Overlap Count: 1
# Description:
#   Bug 1616775 - P1. Add IsThirdPartyWindow and IsThirdPartyTrackingResourceWindow fields to WindowContext. r=timhuang,baku,farre
#   
#   We have to add "IsThirdPartyWindow" in WindowContext because we need to know if
#   a BrowsingContext is third-party (The browsing context may be not
#   in-process).
# ==============================================================================

diff -r 685b68485248 -r ae01b3db6250 dom/base/nsGlobalWindowOuter.cpp
--- a/dom/base/nsGlobalWindowOuter.cpp	Thu Apr 23 04:33:57 2020 +0000
+++ b/dom/base/nsGlobalWindowOuter.cpp	Thu Apr 23 14:24:56 2020 +0000
@@ -2507,6 +2507,9 @@
   ReportLargeAllocStatus();
   mLargeAllocStatus = LargeAllocStatus::NONE;
 
+  bool isThirdPartyTrackingResourceWindow =
+      nsContentUtils::IsThirdPartyTrackingResourceWindow(newInnerWindow);
+
   // Set the cookie jar settings to the window context.
   if (newInnerWindow) {
     net::CookieJarSettingsArgs cookieJarSettings;
@@ -2516,6 +2519,16 @@
     newInnerWindow->GetWindowGlobalChild()
         ->WindowContext()
         ->SetCookieJarSettings(Some(cookieJarSettings));
+
+    newInnerWindow->GetWindowGlobalChild()
+        ->WindowContext()
+        ->SetIsThirdPartyWindow(nsContentUtils::IsThirdPartyWindowOrChannel(
+            newInnerWindow, nullptr, nullptr));
+
+    newInnerWindow->GetWindowGlobalChild()
+        ->WindowContext()
+        ->SetIsThirdPartyTrackingResourceWindow(
+            isThirdPartyTrackingResourceWindow);
   }
 
   mHasStorageAccess = false;
@@ -2539,7 +2552,7 @@
           cookieBehavior == nsICookieService::BEHAVIOR_REJECT_TRACKER ||
           cookieBehavior ==
               nsICookieService::BEHAVIOR_REJECT_TRACKER_AND_PARTITION_FOREIGN);
-      if (nsContentUtils::IsThirdPartyTrackingResourceWindow(newInnerWindow)) {
+      if (isThirdPartyTrackingResourceWindow) {
         checkStorageAccess = true;
       }
     }